cli_tools/gce_vm_image_import/util/scratch_bucket_creator.go

```diff
@@ -0,0 +1,129 @@
+//  Copyright 2019 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package gcevmimageimportutil
+
+import (
+	"cloud.google.com/go/storage"
+	"context"
+	"fmt"
+	"github.com/GoogleCloudPlatform/compute-image-tools/cli_tools/gce_vm_image_import/domain"
+	"google.golang.org/api/iterator"
+	"log"
+	"strings"
+)
+
+const (
+	defaultRegion       = "US"
+	defaultStorageClass = "MULTI_REGIONAL"
+)
+
+// ScratchBucketCreator is responsible for creating Daisy scratch bucketets
+type ScratchBucketCreator struct {
+	StorageClient         domain.StorageClientInterface
+	Ctx                   context.Context
+	BucketIteratorCreator domain.BucketIteratorCreatorInterface
+}
+
+// NewScratchBucketCreator creates a ScratchBucketCreator
+func NewScratchBucketCreator(ctx context.Context, storageClient *storage.Client) *ScratchBucketCreator {
+	return &ScratchBucketCreator{NewStorageClient(ctx, storageClient), ctx, &BucketIteratorCreator{}}
+}
+
+// CreateScratchBucket creates scratch bucket in the same region as sourceFileFlag.
+// Returns (bucket_name, region, error)
+func (c *ScratchBucketCreator) CreateScratchBucket(
+	sourceFileFlag string, project string) (string, string, error) {
+	bucket := ""
+	region := ""
+	var err error
+
+	if project == "" {
+		return "", "", fmt.Errorf("can't create scratch bucket if project not specified")
+	}
+
+	if sourceFileFlag != "" {
+		// source file provided, create bucket in the same region for cost/performance reasons
+		bucket, region, err = c.createBucketMatchFileRegion(sourceFileFlag, project)
+	}
+
+	if sourceFileFlag == "" || err != nil {
+		// source file not provided or couldn't create bucket based on it, create default scratch bucket
```

Comment:
This woudl be more idiomatic:

if sourceFileFlag == "" || err != nil {
  bucket := c.formatScratchBucketName(project, defaultRegion)
  region, err = c.createBucketIfNotExisting(bucket, project, &storage.BucketAttrs{Name: bucket, Location: defaultRegion, StorageClass: defaultStorageClass})
  if err != nil {
    return "", "", err
  }
}
return bucket, region, nil

Reply:
Done